src/modules/map/components/CompareSlider.vue

```diff
@@ -10,44 +11,121 @@ const props = defineProps({
         default: window.innerWidth,
     },
 })
-
+const getMap = inject('getMap')
 const { clientWidth } = toRefs(props)
 
-const compareRatio = ref(-0.5)
+const preComposeKey = ref(null)
+const postComposeKey = ref(null)
+const compareSliderOffset = ref(0)
 
+const compareRatio = ref(-0.5)
 const store = useStore()
 const storeCompareRatio = computed(() => store.state.ui.compareRatio)
 const compareSliderPosition = computed(() => {
     return {
         left: compareRatio.value * 100 + '%',
     }
 })
+const visibleLayerOnTop = computed(() => store.getters.visibleLayerOnTop)
 
 watch(storeCompareRatio, (newValue) => {
     compareRatio.value = newValue
+    slice()
+})
+
+watch(visibleLayerOnTop, () => {
+    nextTick(slice())
 })
 
 onMounted(() => {
     compareRatio.value = storeCompareRatio.value
+    slice()
 })
 
-function grabSlider() {
+function slice() {
+    if (preComposeKey.value != null && postComposeKey.value != null) {
+        unByKey(preComposeKey.value)
+        unByKey(postComposeKey.value)
+        preComposeKey.value = null
+        postComposeKey.value = null
+    }
+
+    if (
+        getMap()
+            ?.getAllLayers()
+            .find((layer) => {
+                return (
+                    visibleLayerOnTop.value && layer.get('id') === visibleLayerOnTop.value.getID()
+                )
+            })
+    ) {
+        preComposeKey.value = getMap()
+            ?.getAllLayers()
+            .find((layer) => {
+                return (
+                    visibleLayerOnTop.value && layer.get('id') === visibleLayerOnTop.value.getID()
+                )
+            })
+            .on('prerender', onPreCompose)
+
+        postComposeKey.value = getMap()
+            ?.getAllLayers()
+            .find((layer) => {
+                return (
+                    visibleLayerOnTop.value && layer.get('id') === visibleLayerOnTop.value.getID()
+                )
+            })
+            .on('postrender', onPostCompose)
+    }
```

Comment:
you should create a variable with the result of this `getMap()?.getAllLayers().find(...)` as you are reusing it twice afterward